nexus/reconfigurator/execution/src/datasets.rs

```diff
@@ -4,97 +4,214 @@
+        let id = bp_dataset.id;
+        let kind = &bp_dataset.kind;
 
-        // If already present in the datastore, move on.
-        if existing_datasets.remove(&id) {
-            num_already_exist += 1;
-            continue;
-        }
+        // If this dataset already exists, only update it if it appears different from what exists
+        // in the database already.
+        let action = if let Some(db_dataset) = existing_datasets.remove(&id) {
+            let db_config: DatasetConfig = db_dataset.try_into()?;
+            let bp_config: DatasetConfig = bp_dataset.clone().try_into()?;
 
-        let pool_id = dataset.pool_name.id();
-        let dataset = Dataset::new(
-            id.into_untyped_uuid(),
-            pool_id.into_untyped_uuid(),
-            Some(address),
-            kind.clone(),
-        );
-        let maybe_inserted = datastore
-            .dataset_insert_if_not_exists(dataset)
-            .await
-            .with_context(|| {
-                format!("failed to insert dataset record for dataset {id}")
-            })?;
-
-        // If we succeeded in inserting, log it; if `maybe_dataset` is `None`,
-        // we must have lost the TOCTOU race described above, and another Nexus
-        // must have inserted this dataset before we could.
-        if maybe_inserted.is_some() {
-            info!(
-                opctx.log,
-                "inserted new dataset for Omicron zone";
-                "id" => %id,
-                "kind" => ?kind,
-            );
-            num_inserted += 1;
+            if db_config == bp_config {
+                num_unchanged += 1;
+                continue;
+            }
+            num_updated += 1;
+            "update"
         } else {
-            num_already_exist += 1;
+            num_inserted += 1;
+            "insert"
+        };
+
+        let dataset = Dataset::from(bp_dataset.clone());
+        datastore.dataset_upsert(dataset).await.with_context(|| {
+            format!("failed to upsert dataset record for dataset {id}")
+        })?;
+
+        info!(
+            opctx.log,
+            "ensuring dataset record in database";
+            "action" => action,
+            "id" => %id,
+            "kind" => ?kind,
+        );
+    }
+
+    for bp_dataset in unwanted_datasets {
+        if existing_datasets.remove(&bp_dataset.id).is_some() {
+            if matches!(
+                bp_dataset.kind,
+                omicron_common::disk::DatasetKind::Crucible
+            ) {
+                // Region and snapshot replacement cannot happen without the
+                // database record, even if the dataset has been expunged.
+                //
+                // This record will still be deleted, but it will happen as a
+                // part of the "decommissioned_disk_cleaner" background task.
+                continue;
+            }
+
+            datastore.dataset_delete(&opctx, bp_dataset.id).await?;
```

Comment:
Now that we're removing datasets from the db, do we need to be concerned about Nexuses executing on older blueprints? E.g.,

* Blueprint A adds dataset X, and is the current target
* Nexus 1 successfully executes blueprint A, adding the dataset record
* Blueprint B expunges dataset X, and is the made the new target
* Nexus 1 successfully executes blueprint B, deleting the dataset record
* Nexus 2 executes blueprint A, and tries to add the dataset record again?

For external networking we ended up putting the entire set of addition/removal into a transaction that was gated on the current blueprint being the target to prevent this kind of thing, but that's a pretty heavy hammer (and I'd still prefer we had a lighter means on that one).

I also wonder if we need to invest in some kind of testing help for this kind of problem. "What if a Nexus executes an older blueprint" seems like it's hard to get my head around _generally_, and I often forget to ask that question as I'm initially writing code in this crate. Seems like we could easily have lurking problems _already_ (e.g., the external networking stuff landed first, then only later we moved it to a transaction when we realized it was broken in this case).

Reply:
(Should we just gate everything on current blueprint = target? If that makes reasoning about the system simpler and the cost isn't excessive then it seems worth it to me.)

Reply:
Just to confirm on the "blueprint = target" stuff -- would that be "reading from the `bp_target` table for the value with the highest version? Or should I be looking for the highest value with `enabled = true`?

Reply:
I'd be happy to take each of the "data modifying" portions of the executor, and have them additionally make them check for "blueprint = target" in a transaction. That would avoid creating a "single large transaction", while also avoiding any modifications if the blueprint changes partway through execution.

Reply:
> Just to confirm on the "blueprint = target" stuff -- would that be "reading from the `bp_target` table for the value with the highest version? Or should I be looking for the highest value with `enabled = true`?

The current target is always "the row in `bp_target` with the highest version". Whether it's enabled is orthogonal, and the value of `enabled` for any row except the one with the highest version is irrelevant.

I think in this context, you already know the blueprint you're trying to execute, and the fact that you're trying to execute at all means it was enabled when this RPW started. I think if we're trying to check "...and this is still the target", it's sufficient to confirm that the ID of the blueprint we're executing matches the highest-versioned ID in `bp_target`. I don't think we need to also check whether it's enabled; that's inherently racy anyway.

Reply:
I added this support in 50019ee95bd274238a35d0abd453ade8c85a5d98 - this was not a trivial operation, so it's probably worth a close look.

Of note:

- I added `transaction_if_current_blueprint_is` to make a generic transaction, which aborts if the blueprint UUID doesn't match what we expect
- I'm using this for dataset upsert/deletion, and added tests
- I updated some old tests to make sure a current blueprint target exists and looks valid